modules/globebrowsing/util/globelabelscomponent.cpp

```diff
@@ -0,0 +1,842 @@
+        invModelMatrix * glm::dvec4(data.camera.viewDirectionWorldSpace(), 0.0)
+    );
+    glm::dvec3 cameraUpDirectionObj = glm::dvec3(
+        invModelMatrix * glm::dvec4(data.camera.lookUpVectorWorldSpace(), 0.0)
+    );
+    glm::dvec3 orthoRight = glm::normalize(
+        glm::cross(cameraViewDirectionObj, cameraUpDirectionObj)
+    );
+    if (orthoRight == glm::dvec3(0.0)) {
+        glm::dvec3 otherVector(
+            cameraUpDirectionObj.y,
+            cameraUpDirectionObj.x,
+            cameraUpDirectionObj.z
+        );
+        orthoRight = glm::normalize(glm::cross(otherVector, cameraViewDirectionObj));
+    }
+    glm::dvec3 orthoUp = glm::normalize(
+        glm::cross(orthoRight, cameraViewDirectionObj)
+    );
+
+    for (const LabelEntry lEntry : _labels.labelsArray) {
+        glm::vec3 position = lEntry.geoPosition;
+        glm::dvec3 locationPositionWorld =
+            glm::dvec3(_globe->modelTransform() * glm::dvec4(position, 1.0));
+        double distanceCameraToLabelWorld =
+            glm::length(locationPositionWorld - data.camera.positionVec3());
+
+        bool draw = false;
+        if (_labelsDisableCullingEnabled) {
+            draw = true;
+        }
+        else if ((distToCamera > (distanceCameraToLabelWorld + _labelsDistaneEPS)) &&
+            isLabelInFrustum(VP, locationPositionWorld)) { // culling
+            draw = true;
+        }
+
+        if (_labelAlignmentOption == 1) {
+            glm::dvec3 labelNormalObj = glm::dvec3(
+                invModelMatrix * glm::dvec4(data.camera.positionVec3(), 1.0)
+            ) - glm::dvec3(position);
+
+            glm::dvec3 labelUpDirectionObj = glm::dvec3(position);
+            
+            orthoRight = glm::normalize(
+                glm::cross(labelUpDirectionObj, labelNormalObj)
+            );
+            if (orthoRight == glm::dvec3(0.0)) {
+                glm::dvec3 otherVector(
+                    labelUpDirectionObj.y,
+                    labelUpDirectionObj.x,
+                    labelUpDirectionObj.z
+                );
+                orthoRight = glm::normalize(glm::cross(otherVector, labelNormalObj));
+            }
+            orthoUp = glm::normalize(
+                glm::cross(labelNormalObj, orthoRight)
+            );
+        }
+
+        if (draw) {
+            position += _labelsMinHeight;
+            ghoul::fontrendering::FontRenderer::defaultProjectionRenderer().render(
+                *_font,
+                position,
+                lEntry.feature,
+                textColor,
+                powf(2.f, _labelsSize),
+                _labelsMinSize,
+                _labelsMaxSize,
+                modelViewProjectionMatrix,
+                orthoRight,
+                orthoUp,
+                cameraPosObj,
+                cameraLookUpObj,
+                0
+            );
+        }
+    }
+}
+
+bool GlobeLabelsComponent::isLabelInFrustum(const glm::dmat4& MVMatrix,
```

Comment:
Can this be moved out to a general utility for frustum culling, that others could use?

Reply:
Yes, it can and it will. :-)